File: src/main/java/net/pterodactylus/sone/freenet/wot/IdentityManagerImpl.java
/*
 * Sone - IdentityManagerImpl.java - Copyright © 2010–2016 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.freenet.wot;

import static java.util.logging.Logger.getLogger;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import net.pterodactylus.sone.freenet.plugin.PluginException;
import net.pterodactylus.util.service.AbstractService;

import com.google.common.collect.Sets;
import com.google.common.eventbus.EventBus;
import com.google.inject.Inject;
import com.google.inject.Singleton;

/**
 * The identity manager takes care of loading and storing identities, their
 * contexts, and properties. It does so in a way that does not expose errors via
 * exceptions but it only logs them and tries to return sensible defaults.
 * <p>
 * It is also responsible for polling identities from the Web of Trust plugin
 * and sending events to the {@link EventBus} when {@link Identity}s and
 * {@link OwnIdentity}s are discovered or disappearing.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
@Singleton
public class IdentityManagerImpl extends AbstractService implements IdentityManager {

	/** The logger. */
	private static final Logger logger = getLogger(IdentityManagerImpl.class.getName());

	/** The event bus. */
	private final EventBus eventBus;

	private final IdentityLoader identityLoader;

	/** The Web of Trust connector. */
	private final WebOfTrustConnector webOfTrustConnector;

	/** The currently known own identities. */
	private final Set<OwnIdentity> currentOwnIdentities = Sets.newHashSet();

	/**
	 * Creates a new identity manager.
	 *
	 * @param eventBus
	 *            The event bus
	 * @param webOfTrustConnector
	 *            The Web of Trust connector
	 */
	@Inject
	public IdentityManagerImpl(EventBus eventBus, WebOfTrustConnector webOfTrustConnector, IdentityLoader identityLoader) {
		super("Sone Identity Manager", false);
		this.eventBus = eventBus;
		this.webOfTrustConnector = webOfTrustConnector;
		this.identityLoader = identityLoader;
	}

	//
	// ACCESSORS
	//

	/**
	 * Returns whether the Web of Trust plugin could be reached during the last
	 * try.
	 *
	 * @return {@code true} if the Web of Trust plugin is connected,
	 *         {@code false} otherwise
	 */
	@Override
	public boolean isConnected() {
		try {
			webOfTrustConnector.ping();
			return true;
		} catch (PluginException pe1) {
			/* not connected, ignore. */
			return false;
		}
	}

	/**
	 * Returns all own identities.
	 *
	 * @return All own identities
	 */
	@Override
	public Set<OwnIdentity> getAllOwnIdentities() {
		synchronized (currentOwnIdentities) {
			return new HashSet<OwnIdentity>(currentOwnIdentities);
		}
	}

	//
	// SERVICE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected void serviceRun() {
		Map<OwnIdentity, Collection<Identity>> oldIdentities = new HashMap<OwnIdentity, Collection<Identity>>();

		while (!shouldStop()) {
			try {
				Map<OwnIdentity, Collection<Identity>> currentIdentities = identityLoader.loadIdentities();

				IdentityChangeEventSender identityChangeEventSender = new IdentityChangeEventSender(eventBus, oldIdentities);
				identityChangeEventSender.detectChanges(currentIdentities);

				oldIdentities = currentIdentities;

				synchronized (currentOwnIdentities) {
					currentOwnIdentities.clear();
					currentOwnIdentities.addAll(currentIdentities.keySet());
				}
			} catch (WebOfTrustException wote1) {
				logger.log(Level.WARNING, "WoT has disappeared!", wote1);
			}

			/* wait 15 minutes before checking again. */
			sleep(15 * 60 * 1000);
		}
	}

}

File: src/main/kotlin/net/pterodactylus/sone/freenet/wot/IdentityManagerImpl.kt
/*
 * Sone - IdentityManagerImpl.kt - Copyright © 2010–2020 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.freenet.wot

import com.google.common.eventbus.EventBus
import com.google.common.eventbus.Subscribe
import com.google.inject.Inject
import com.google.inject.Singleton
import net.pterodactylus.sone.core.event.StrictFilteringActivatedEvent
import net.pterodactylus.sone.core.event.StrictFilteringDeactivatedEvent
import net.pterodactylus.util.service.AbstractService
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicBoolean
import java.util.logging.Level
import java.util.logging.Logger
import java.util.logging.Logger.getLogger

/**
 * The identity manager takes care of loading and storing identities, their
 * contexts, and properties. It does so in a way that does not expose errors via
 * exceptions but it only logs them and tries to return sensible defaults.
 *
 *
 * It is also responsible for polling identities from the Web of Trust plugin
 * and sending events to the [EventBus] when [Identity]s and
 * [OwnIdentity]s are discovered or disappearing.
 */
@Singleton
class IdentityManagerImpl @Inject constructor(
		private val eventBus: EventBus,
		private val webOfTrustConnector: WebOfTrustConnector,
		private val identityLoader: IdentityLoader
) : AbstractService("Sone Identity Manager", false), IdentityManager {

	private val currentOwnIdentities = mutableSetOf<OwnIdentity>()
	private val strictFiltering = AtomicBoolean(false)

	override val isConnected: Boolean
		get() = notThrowing { webOfTrustConnector.ping() }

	override val allOwnIdentities: Set<OwnIdentity>
		get() = synchronized(currentOwnIdentities) {
			currentOwnIdentities.toSet()
		}

	override fun serviceRun() {
		var oldIdentities = mapOf<OwnIdentity, Collection<Identity>>()

		while (!shouldStop()) {
			try {
				val currentIdentities = identityLoader.loadAllIdentities().applyStrictFiltering()

				val identityChangeEventSender = IdentityChangeEventSender(eventBus, oldIdentities)
				identityChangeEventSender.detectChanges(currentIdentities)

				oldIdentities = currentIdentities

				synchronized(currentOwnIdentities) {
					currentOwnIdentities.clear()
					currentOwnIdentities.addAll(currentIdentities.keys)
				}
			} catch (wote1: WebOfTrustException) {
				logger.log(Level.WARNING, "WoT has disappeared!", wote1)
			} catch (e: Exception) {
				logger.log(Level.SEVERE, "Uncaught exception in IdentityManager thread!", e)
			}

			/* wait a minute before checking again. */
			sleep(SECONDS.toMillis(60))
		}
	}

	private fun Map<OwnIdentity, Set<Identity>>.applyStrictFiltering() =
			if (strictFiltering.get()) {
				val identitiesWithTrust = values.flatten()
						.groupBy { it.id }
						.mapValues { (_, identities) ->
							identities.reduce { accIdentity, identity ->
								identity.trust.forEach { (ownIdentity: OwnIdentity?, trust: Trust?) ->
									accIdentity.setTrust(ownIdentity, trust)
								}
								accIdentity
							}
						}

				mapValues { (_, trustedIdentities) ->
					trustedIdentities.filter { trustedIdentity ->
						identitiesWithTrust[trustedIdentity.id]!!.trust.all { it.value.hasZeroOrPositiveTrust() }
					}
				}
			} else {
				this
			}

	@Subscribe
	fun strictFilteringActivated(event: StrictFilteringActivatedEvent) {
		strictFiltering.set(true)
	}

	@Subscribe
	fun strictFilteringDeactivated(event: StrictFilteringDeactivatedEvent) {
		strictFiltering.set(false)
	}

}

private val logger: Logger = getLogger(IdentityManagerImpl::class.java.name)

private fun notThrowing(action: () -> Unit): Boolean =
		try {
			action()
			true
		} catch (e: Exception) {
			false
		}

private fun Trust.hasZeroOrPositiveTrust() =
		if (explicit == null) {
			implicit == null || implicit >= 0
		} else {
			explicit >= 0
		}
